File: solutions/227.basic-calculator-ii.js
/*
 * @lc app=leetcode id=227 lang=javascript
 *
 * [227] Basic Calculator II
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var calculate = function (s) {
  s = s.replaceAll(" ", "");

  const map = new Map([
    ["*", 2],
    ["/", 2],
    ["+", 1],
    ["-", 1],
  ]);

  const ops = [],
    nums = [];

  // 清算函数
  const calc = () => {
    if (nums.length < 2) return;
    const op = ops.pop();
    const n2 = nums.pop(),
      n1 = nums.pop();
    let res;
    switch (op) {
      case "*":
        res = n1 * n2;
        break;
      case "/":
        res = Math.floor(n1 / n2);
        break;
      case "+":
        res = n1 + n2;
        break;
      case "-":
        res = n1 - n2;
        break;
    }
    nums.push(res);
  };

  for (let i = 0; i < s.length; ++i) {
    const c = s[i];
    if (isNaN(c)) {
      // + - * /
      if (i === 0) nums.push(0); // add 0: -1 -> 0-1
      while (map.get(c) <= map.get(ops.at(-1))) calc(); // current <= 栈内优先级, 清算, eg 2*3 遇到+4 则清算 2*3
      ops.push(c);
    } else {
      // 数字
      let n = 0;
      let j = i;
      while (j < s.length && !isNaN(s[j])) n = n * 10 + Number(s[j++]);
      i = j - 1;
      nums.push(n);
    }
  }
  while (ops.length) calc();
  return nums.at(-1);
};
// @lc code=end

File: solutions/227.basic-calculator-ii.rb
# @param {String} s
# @return {Integer}
def calculate(s)
  nums = []
  ops = []
  priority = { '+' => 1, '-' => 1, '*' => 2, '/' => 2 }
  i = 0
  while i < s.size
    c = s[i]
    case c
    when /\d/
      num = 0
      while i < s.size && s[i].match(/\d/)
        num = num * 10 + s[i].to_i
        i += 1
      end
      nums << num
      i -= 1
    when '+', '-', '*', '/'
      nums << basic_calc(ops.pop, *nums.pop(2)) until ops.empty? || priority[ops.last] < priority[c]
      ops << c
    end
    i += 1
  end
  nums << basic_calc(ops.pop, *nums.pop(2)) until ops.empty?
  nums[0]
end

def basic_calc(op, a, b)
  case op
  when '+'
    a + b
  when '-'
    a - b
  when '*'
    a * b
  when '/'
    a / b
  end
end
